client/challenge.go

```diff
@@ -14,33 +14,50 @@ import (
 	"github.com/multiformats/go-multiaddr"
 )
 
-// SendChallenge submits a challenge to the DNS server for the given peerID.
+// SendChallenge submits value for DNS-01 challenge to the p2p-forge HTTP server for the given peerID.
 // It requires the corresponding private key and a list of multiaddresses that the peerID is listening on using
 // publicly reachable IP addresses.
-func SendChallenge(ctx context.Context, baseURL string, privKey crypto.PrivKey, challenge string, addrs []multiaddr.Multiaddr) error {
-	req, err := ChallengeRequest(ctx, baseURL, challenge, addrs)
+func SendChallenge(ctx context.Context, baseURL string, privKey crypto.PrivKey, challenge string, addrs []multiaddr.Multiaddr, forgeAuth string, userAgent string, modifyForgeRequest func(r *http.Request) error) error {
+	// Create request
+	registrationURL := fmt.Sprintf("%s/v1/_acme-challenge", baseURL)
+	req, err := ChallengeRequest(ctx, registrationURL, challenge, addrs)
 	if err != nil {
 		return err
 	}
 
+	// Adjust headers if needed
+	if forgeAuth != "" {
+		req.Header.Set(ForgeAuthHeader, forgeAuth)
+	}
+	if userAgent == "" {
+		userAgent = defaultUserAgent
+	}
+	req.Header.Set("User-Agent", userAgent)
+	if modifyForgeRequest != nil {
+		if err := modifyForgeRequest(req); err != nil {
+			return err
+		}
+	}
+
+	// Execute request wrapped in ClientPeerIDAuth
 	client := &httppeeridauth.ClientPeerIDAuth{PrivKey: privKey}
 	_, resp, err := client.AuthenticatedDo(http.DefaultClient, req)
 	if err != nil {
-		return err
+		return fmt.Errorf("libp2p HTTP ClientPeerIDAuth error at %s: %w", registrationURL, err)
 	}
 	if resp.StatusCode != http.StatusOK {
 		respBody, _ := io.ReadAll(resp.Body)
-		return fmt.Errorf("%s : %s", resp.Status, respBody)
+		return fmt.Errorf("%s error from %s: %q", resp.Status, registrationURL, respBody)
 	}
 	return nil
 }
 
-// ChallengeRequest creates an HTTP Request object for submitting an ACME challenge to the DNS server for a given peerID.
+// ChallengeRequest creates an HTTP Request object for submitting an ACME challenge to the p2p-forge HTTP server for a given peerID.
 // Construction of the request requires a list of multiaddresses that the peerID is listening on using
 // publicly reachable IP addresses.
 //
 // Sending the request to the DNS server requires performing HTTP PeerID Authentication for the corresponding peerID
-func ChallengeRequest(ctx context.Context, baseURL string, challenge string, addrs []multiaddr.Multiaddr) (*http.Request, error) {
+func ChallengeRequest(ctx context.Context, registrationURL string, challenge string, addrs []multiaddr.Multiaddr) (*http.Request, error) {
```

Comment:
@lidel the reason why the test is now failing is that you changed the semantics to this function which is the one called by the test (i.e. there is now a 404 here because adding "/v1/_acme-challenge" was removed.  So the test needs modification also (e.g. to the changed ChallengeRequest semantics or perhaps to just use SendChallenge now).

It might also be worth considering if both of these functions need exporting or not.

Reply:
Thanks! 

Let's keep them, I feel they may be useful in testing, solving the problem of the same code in 2-3 places, like in was in our case before this PR.